### PR TITLE
UnderlineNav: Add `overflow: hidden` when calculating items for overflow menu

### PR DESCRIPTION
**Pull request by TylerJDev**
Closes https://github.com/github/primer/issues/6291

Credits to @siddharthkp & @francinelucca for the original PR: https://github.com/primer/react/pull/7463 ✨  

#### Changed

* Adds `overflow: hidden` to container when loading items

### Rollout strategy

- [x] Patch release
- [ ] Minor release
- [ ] Major release; if selected, include a written rollout or migration plan
- [ ] None; if selected, include a brief description as to why

### Merge checklist

- [ ] Added/updated tests
- [ ] Added/updated documentation
- [ ] Added/updated previews (Storybook)
- [x] Changes are [SSR compatible](https://github.com/primer/react/blob/main/contributor-docs/CONTRIBUTING.md#ssr-compatibility)
- [x] Tested in Chrome
- [ ] Tested in Firefox
- [ ] Tested in Safari
- [ ] Tested in Edge
- [ ] (GitHub staff only) Integration tests pass at github/github ([Learn more about how to run integration tests](https://github.com/github/primer-engineering/blob/main/how-we-work/testing-primer-react-pr-at-dotcom.md))

### CODE REVIEW

File: packages/react/src/internal/components/UnderlineTabbedInterface.module.css [hunk 1]
```diff
@@ -12,6 +12,13 @@
   /* stylelint-disable-next-line primer/box-shadow */
   box-shadow: inset 0 -1px var(--borderColor-muted);
 
+  /* Hide overflow until calculation is complete to prevent CLS */
+  overflow: hidden;
+
```
[the remark this answers is not in the document]
Comment (TylerJDev) [reply packages/react/src/internal/components/UnderlineTabbedInterface.module.css]: I think this is fair!

File: packages/react/src/internal/components/UnderlineTabbedInterface.module.css [hunk 2]
```diff
@@ -12,6 +12,16 @@
   /* stylelint-disable-next-line primer/box-shadow */
   box-shadow: inset 0 -1px var(--borderColor-muted);
 
+  /* Hide overflow until calculation is complete to prevent CLS */
+  overflow: visible;
```
Comment (adierkens) [inline packages/react/src/internal/components/UnderlineTabbedInterface.module.css:16]: I think we can do this entirely within css and avoid any of the `useEffect()` updates, just with something like: 

```css
.UnderlineWrapper {
  &:not(:has([class*='prc-UnderlineNav-MoreButton-'])) {
    overflow-x: hidden;
  }
}
```

You might need to update the classNames a bit (or move it to the `UnderlineNav` component directly)

Comment (siddharthkp) [reply packages/react/src/internal/components/UnderlineTabbedInterface.module.css:16]: Very nice! Although probably the same thing perf wise because the MoreButton appears after a javascript pass as well, right?

Thinking out loud, do we need to test this for cases where there is an overflow because of icons and counters, so the MoreButton may never appear, or is that already covered?

File: packages/react/src/internal/components/UnderlineTabbedInterface.tsx [hunk 3]
```diff
@@ -32,14 +32,18 @@ type UnderlineWrapperProps<As extends React.ElementType> = {
...
 
 export const UnderlineWrapper = forwardRef((props, ref) => {
-  const {children, className, as: Component = 'div', ...rest} = props
+  const {children, className, as: Component = 'div', ready, ...rest} = props
+
   return (
     <Component
       className={clsx(classes.UnderlineWrapper, className)}
       ref={ref as ForwardedRef<HTMLDivElement>}
+      data-ready={ready === undefined ? undefined : ready ? 'true' : 'false'}
```
Comment (siddharthkp) [inline packages/react/src/internal/components/UnderlineTabbedInterface.tsx]: More descriptive name please. `data-hydrated` / `data-overflow-handled`?

**Review by siddharthkp — APPROVED** [review]
Left a few comments, approving in advance

File: .changeset/clever-geese-cover.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+'@primer/react': patch
+---
+
+UnderlineNav: Adds `overflow: hidden` when calculating items to prevent CLS
```
Comment (siddharthkp) [inline .changeset/clever-geese-cover.md]:
```suggestion
UnderlineNav: Adds `overflow: hidden` during SSR prevent CLS, removes it after hydration on client.
```

File: packages/react/src/internal/components/UnderlineTabbedInterface.module.css (hunk 2 above)
Comment (TylerJDev) [reply packages/react/src/internal/components/UnderlineTabbedInterface.module.css:16]: > Thinking out loud, do we need to test this for cases where there is an overflow because of icons and counters, so the MoreButton may never appear,

Can the component still overflow without the menu? I haven't seen that, as I assumed that we also took those into account when determining whether to show the menu or not.

**Review by liuliu-dev — APPROVED** [review]
🎉

**Review by siddharthkp — APPROVED** [review]

File: packages/react/src/UnderlineNav/UnderlineNav.tsx [hunk 5]
```diff
@@ -245,10 +247,19 @@ export const UnderlineNav = forwardRef(
...
-    }, [])
+    const updateListAndMenu = useCallback(
+      (props: ResponsiveProps, displayIcons: boolean) => {
+        setResponsiveProps(props)
+        setIconsVisible(displayIcons)
+
+        // Only mark as ready once widths have been measured for all valid children
+        const widths = displayIcons ? childWidthArray : noIconChildWidthArray
+        if (!isHydrated && widths.length > 0 && widths.length === validChildren.length) {
+          setIsHydrated(true)
```
Comment (siddharthkp) [inline packages/react/src/UnderlineNav/UnderlineNav.tsx]: Oh if it's not just hydration but also calculations, maybe we should call it `data-overflow-measured` or something similar?

File: packages/react/src/UnderlineNav/UnderlineNav.tsx [hunk 6]
```diff
@@ -330,7 +341,14 @@ export const UnderlineNav = forwardRef(
...
       >
         {ariaLabel && <VisuallyHidden as="h2">{`${ariaLabel} navigation`}</VisuallyHidden>}
-        <UnderlineWrapper as={as} aria-label={ariaLabel} className={className} ref={navRef} data-variant={variant}>
+        <UnderlineWrapper
+          as={as}
+          aria-label={ariaLabel}
+          className={className}
+          ref={navRef}
+          data-variant={variant}
+          hydrated={isHydrated}
```
Comment (siddharthkp) [inline packages/react/src/UnderlineNav/UnderlineNav.tsx]: Can we pass `data-hydrated={isHydrated}` and let it spread. Not a fan of the extra prop

**Comment by TylerJDev** [comment]
One more review @siddharthkp 🙏 I updated some of the logic, as the data-* attribute wasn't getting set ([as mentioned by @iansan5653](https://github.com/primer/react/pull/7504#pullrequestreview-3764272476))

**Review by siddharthkp — APPROVED** [review]

**Comment by TylerJDev** [comment]
[VRT passes](https://github.com/github/github-ui/actions/runs/22129308917), skipping integration tests as it seems to be stuck.
